src/esp/core/logging.h

```diff
@@ -7,6 +7,117 @@
 
 #include "esp/core/configure.h"
 
+#include <Corrade/Containers/Array.h>
+#include <Corrade/Containers/StringView.h>
+#include <Corrade/Utility/Debug.h>
+
+namespace esp {
+namespace logging {
+// These are all subsystems that have different logging levels
+// They correspond to namespaces.  When adding a new subsystem, make sure
+// to add an appropriate loggingSubsystem function (see the ADD_SUBSYSTEM_FN
+// helper bellow) and add it to @ref subsystemFromName()
+enum class Subsystem : uint8_t {
+  Gfx,
+  Scene,
+  Sim,
+  Physics,
+  // This is the catch all subsystem
+  Other,
+
+  // Must always be last
+  NumSubsystems,
+};
+
+Subsystem subsystemFromName(Corrade::Containers::StringView name);
+
+}  // namespace logging
+
+// This is the catch all for subsystems without a specialization
+inline logging::Subsystem loggingSubsystem() {
+  return logging::Subsystem::Other;
+}
+
+#define ADD_SUBSYSTEM_FN(subsystemNamespace, subsystemName) \
+  namespace subsystemNamespace {                            \
+  inline logging::Subsystem loggingSubsystem() {            \
+    return logging::Subsystem::subsystemName;               \
+  }                                                         \
+  }
+
+ADD_SUBSYSTEM_FN(gfx, Gfx);
+ADD_SUBSYSTEM_FN(scene, Scene);
+ADD_SUBSYSTEM_FN(sim, Sim);
+ADD_SUBSYSTEM_FN(physics, Physics);
+
+namespace logging {
+
+enum class LoggingLevel : uint8_t {
+  // Always first
+  verbose,
+  debug,
+  warning,
+  // Quiet mode disables debug and warning
+  quiet,
+  error,
+};
```

Comment:
The intent here is that verbose always gives you everything and quiet always silences everything besides errors or worse.  That way we can add new levels without changing things downstream.  verbose and quiet are the two I expect users to use while the others may be useful in specific circumstances.

Reply:
In other words, just so I understand correctly, the following would be functionally equivalent, right?

```cpp
enum class LoggingLevel : uint8_t {
  debug,
  warning,
  error,
  
  // Verbose is debug and all higher levels
  verbose = debug,
  
  // Quiet is only errors and nothing else
  quiet = error
};
```

Reply:
This is excellent! A few questions/comments

1. I would like to see an "info" level - debug is, to me, a loaded term.  An example of an "info"-level message would be some output of a successful transaction that would have resulted in a warning/error otherwise, for example - perhaps not as detailed as "debug" info, and probably not used in the same contexts (with "debug" documenting potential errors vs "info" giving non-negative feedback).
2. How hard would it be to add subsystem-level/logging-level stream routing (i.e. send certain subsystem and/or log-level output to certain log files instead of the console)? I know this would probably be beyond the scope of this PR.  I just would think something like this would be useful by giving folks a way to check on continuing processes without having to search the console output.

Reply:
@mosra, yeah, those are equivalent.  I'll change it to look like that as that makes it clearer.

Reply:
@jturner65 I am currently matching Magnum names that what it is clear which of the Corrade's loggers gets returned by the macro (since they can be configured differently).  Given @mosra's point that even turned off logging isn't free, we should be wary of adding a level that is turned off by default to not inadvertently tank perform due to adding lots of logging statements.

It should be pretty simple to route to a file.  Corrade's loggers just take an ostream point, which can easily be a file.

Reply:
> Given @mosra's point that even turned off logging isn't free, we should be wary of adding a level that is turned off by default to not inadvertently tank perform due to adding lots of logging statements.

That's a valid point. Do we want to add compiler directives around non-warning/error log messages, so they are not even compiled if we do not wish to support them?  or is that overkill do you think?

Reply:
@jturner65 we figured out a way how to make them ~free, after all, with some *cursed* code :)

Reply:
See `_ESP_LOG_IF` for the craziness :)